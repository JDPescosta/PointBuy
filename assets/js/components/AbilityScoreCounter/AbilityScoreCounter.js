import React, { useState, useEffect } from "react";
import IconButton from "../IconButton/IconButton";
import PlusIcon from "../../../images/icons/plusIcon.svg";
import MinusIcon from "../../../images/icons/minusIcon.svg";
import Chevron from "../../../images/icons/chevron-up.svg";
import DoubleChevron from "../../../images/icons/double-chevron-up.svg";
import ChevronOutline from "../../../images/icons/chevron-outline.svg";

import "./AbilityScoreCounter.scss";

const AbilityScoreCounter = ({
  attribute = {},
  racialBonus,
  dynamicScore,
  setDynamicScore,
}) => {
  const [totalScore, setTotalScore] =
    racialBonus === "dynamic" ? useState(8) : useState(8 + racialBonus);
  const [dynamicFlag, setDynamicFlag] = useState(false);
  const [displayBonus, setDisplayBonus] = useState(-1);
  const [tooltipShown, setTooltipShown] = useState(false);

  useEffect(() => {
    let dynamicBonus = +dynamicFlag;

    setTotalScore(
      racialBonus === "dynamic"
        ? attribute.abilityScore + dynamicBonus
        : attribute.abilityScore + racialBonus
    );
  }, [attribute.pointCost, racialBonus, dynamicFlag]);

  useEffect(() => {
    setDisplayBonus(genDisplayBonus(totalScore));
  }, [totalScore]);

  const genDisplayBonus = (score) =>
    Math.floor((totalScore - 10) / 2) > 0
      ? `+ ${Math.floor((totalScore - 10) / 2)}`
      : Math.floor((totalScore - 10) / 2);

  const attText = {
    strength: {
      text: [
        { label: "Carry weight:", value: `${totalScore * 15}lb` },
        { label: "Push weight:", value: `${totalScore * 30}lb` },
        { label: "Long jump:", value: `${totalScore / 2}/${totalScore}ft` },
        {
          label: "High jump:",
          value: `${(3 + (totalScore - 10) / 2) / 2}/${
            3 + (totalScore - 10) / 2
          }ft`,
        },
        { label: "Skills:", value: "Athletics" },
      ],
    },
    dexterity: {
      text: [
        {
          label: "Base initiative:",
          value: displayBonus,
        },
        { label: "Skills:", value: "Acrobatics, Sleight of hand, Stealth" },
      ],
    },
    constitution: {
      text: [
        {
          label: "Hitpoint Bonus:",
          value: `Bonus hitpoints gained each level:  
          ${racialBonus === 2 ? `+ ${Math.floor((totalScore - 10) / 2) + 1}` : displayBonus}`, //This takes into account Hill Dwarves racial of +1 to hp a level 
        },
      ],
    },
    intelligence: {
      text: [
        {
          label: "Skills:",
          value: "Arcana, History, Investigation, Nature, Religion",
        },
      ],
    },
    wisdom: {
      text: [
        {
          label: "Skills:",
          value: "Animal handling, Insight, Medicine, Perception, Survival",
        },
      ],
    },
    charisma: {
      text: [
        {
          label: "Skills:",
          value: "Deception, Intimidation, Performance, Persuasion",
        },
      ],
    },
  };

  const modifyDynamicScore = () => {
    if (!dynamicFlag && dynamicScore > 1) return;

    let modifier = dynamicFlag ? -1 : 1;

    setDynamicScore(dynamicScore + modifier);
    setDynamicFlag(!dynamicFlag);
  };

  const modifyAbilityScore = (isPositive) => {
    let score = attribute.abilityScore;

    if (score < 15 && isPositive) {
      score++;
    } else if (score > 8 && !isPositive) {
      score--;
    }

    let pointCost = score < 14 ? score - 8 : 7 + (score - 14) * 2;

    attribute.setPointCost(pointCost);
    attribute.setAbilityScore(score);
  };

  const ChevronBonus = ({racialBonus, dynamicFlag, dynamicScore}) => (
    <div onMouseEnter={() => setTooltipShown(true)} onMouseLeave={() => setTooltipShown(false)}>
      {racialBonus === "dynamic" && (dynamicFlag || dynamicScore < 2) && (
        <span className={"chevron-anchor"}>
          {tooltipShown && (
            <div className="chevron-tooltip">
              {dynamicFlag && "You've increased this Ability Score by 1."}
              {!dynamicFlag && "You can choose to increase this ability score due to your Ability Score Increase racial."}
            </div>
          )}
          <IconButton
            tabCheck={false}
            isDisabled={false}
            onClick={modifyDynamicScore}
          >
            {dynamicFlag && <Chevron className="chevron" />}
            {!dynamicFlag && <ChevronOutline className="chevron" />}
          </IconButton>
        </span>
      )}
      {racialBonus > 0 && (
        <span className={"chevron-anchor"}>
          {tooltipShown && (
            <div className="chevron-tooltip">Your race is increaing this ability score by {racialBonus}.</div>
          )}
          {racialBonus === 1 ? <Chevron className="chevron" /> : <DoubleChevron className="chevron" />}
        </span>
      )}
    </div>
  );

  return (
    <div className="ability-score-counter">
      <div className="counter-container">
        <div className="score-display">
          <IconButton
            tabCheck={false}
            isDisabled={attribute.abilityScore <= 8}
            onClick={() => modifyAbilityScore(0)}
          >
            <MinusIcon />
          </IconButton>
          <div>{totalScore}</div>
          <IconButton
            tabCheck={false}
            isDisabled={attribute.abilityScore >= 15}
            onClick={() => modifyAbilityScore(1)}
          >
            <PlusIcon />
          </IconButton>
        </div>
        <div className="display-bonus">{displayBonus}</div>
      </div>
      <h2 className="attribute">
        {attribute.name}
        <ChevronBonus racialBonus={racialBonus} dynamicFlag={dynamicFlag} dynamicScore={dynamicScore}/>
      </h2>

      {attText[attribute.name].text.map(({ label, value }, index) => (
        <p key={attribute.name + index}>
          <span>{label}</span> {value}
        </p>
      ))}
    </div>
  );
};

export default AbilityScoreCounter;
