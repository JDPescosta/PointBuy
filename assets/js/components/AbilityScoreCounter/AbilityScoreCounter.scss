@import "../../../scss/variables";

.ability-score-counter {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-top: 24px;
  width: 30%;
  height: 45%;
  min-width: 155px;

  .counter-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    position: relative;

    .score-display {
      display: flex;
      flex-direction: row;
      align-items: center;
      justify-content: center;
      background-color: $background-highlight-color;
      border-radius: 100%;

      border: solid 4px white;

      width: 115.69px;
      height: 115.69px;

      div {
        color: $text-color;
        font-family: $text-font;
        font-size: 36px;
        font-weight: 700;
        line-height: 46px;
        min-width: 40px;
        text-align: center;
      }

      button {
        position: relative;

        display: flex;
        align-items: space-between;

        justify-content: center;

        background-color: transparent;
        border: none;
        border-radius: 100%;

        color: $text-color;

        font-family: $text-font;
        font-size: 30px;
        font-weight: 700;
        line-height: 30px;
        text-align: center;

        margin: 0 20px 0 20px;

        min-width: 37px;
        height: 37px;
        border-radius: 100%;
        background-color: white;
      }
      button:disabled {
        visibility: hidden;
      }
      button:hover{
        -webkit-box-shadow: 0 0 2px 2px #ccc;  
        -moz-box-shadow:    0 0 2px 2px #ccc;  
        box-shadow:         0 0 2px 2px #ccc;
      }
      button svg {
        fill: $background-highlight-color;
      }
    }

    .display-bonus {
      position: absolute;
      bottom: 0;
      transform: translateY(30%);

      background-color: #1c1c1c;
      border: 2px solid white;
      border-radius: 10px;
      width: 50px;
      height: 30px;

      color: $text-color;
      font-family: $text-font;
      font-size: 23px;
      font-weight: 700;
      text-align: center;

      justify-content: center;
      align-items: center;
    }
  }

  .attribute {
    color: $text-color;
    font-family: $heading-font;
    font-size: 28px;
    font-weight: 400;
    letter-spacing: 1px;
    line-height: 33px;
    width: 183px;
    text-align: center;
    text-transform: capitalize;

    margin: 20px 0 10px 0;

    display: flex;
    flex-direction: row;
    justify-content: center;

    .chevron-anchor {
      width: 32px;
      position: relative;
      .chevron-tooltip {
        font: 14px $text-font;
        color: white;
        background-color: #1c1c1c;
        position: absolute;
        display: inline-block;
        top: 30px;
        right: -25px;
        border-radius: 10px;
        width: 100px;
        padding: 15px;
        text-transform: none;
      }
      .chevron-tooltip::after {
        content: " ";
        position: absolute;
        bottom: 100%;
        left: 90%;
        margin-left: -5px;
        border-width: 5px;
        border-style: solid;
        border-color: transparent transparent #1c1c1c transparent;
      }
      .chevron {
        position: absolute;
        top: -20%;
        left: 10%;
        fill: none;
        width: 32px;
        height: 32px;
      }

      button {
        background-color: transparent;
        border: none;
        width: 32px;
        height: 32px;

        .chevron {
          position: absolute;
          top: -20%;
          left: 10%;
          fill: none;
          width: 32px;
          height: 32px;
        }
      }
    }
  }

  p {
    color: $text-color;
    font-family: $text-font;
    font-size: 18px;
    font-weight: 500;
    line-height: 30px;
    letter-spacing: 1px;
    width: 183px;
    text-align: center;
    margin: 0;

    span {
      font-family: $heading-font;
      font-size: 20px;
      font-weight: 500;
      letter-spacing: 1px;
    }
  }
}
