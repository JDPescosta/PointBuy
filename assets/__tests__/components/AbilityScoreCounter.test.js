import React from "react";
import AbilityScoreCounter from "../../js/components/AbilityScoreCounter/AbilityScoreCounter";
import {render} from "@testing-library/react";

test("Renders AbilityScoreCounter", () => {
  let abilityScoreCounter = render(<AbilityScoreCounter attribute={{name: "strength", pointCost: 0, abilityScore: 0, setPointCost: () => {}, setAbilityScore: () => {}}}/>)

  expect(abilityScoreCounter).toMatchSnapshot();
})

