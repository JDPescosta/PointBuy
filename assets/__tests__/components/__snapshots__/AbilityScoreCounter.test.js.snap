// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders AbilityScoreCounter 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ability-score-counter"
      >
        <div
          class="counter-container"
        >
          <div
            class="score-display"
          >
            <button
              class="button-container is-disabled"
              disabled=""
              type="button"
            >
              <test-file-stub />
            </button>
            <div>
              NaN
            </div>
            <button
              class="button-container"
              type="button"
            >
              <test-file-stub />
            </button>
          </div>
          <div
            class="bonus-display"
          >
            NaN
          </div>
        </div>
        <h2
          class="attribute"
        >
          strength
        </h2>
        <p>
          <span>
            Carry weight:
          </span>
           
          NaNlb
        </p>
        <p>
          <span>
            Push weight:
          </span>
           
          NaNlb
        </p>
        <p>
          <span>
            Long jump:
          </span>
           
          NaN/NaNft
        </p>
        <p>
          <span>
            High jump:
          </span>
           
          NaN/NaNft
        </p>
        <p>
          <span>
            Skills:
          </span>
           
          Athletics
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ability-score-counter"
    >
      <div
        class="counter-container"
      >
        <div
          class="score-display"
        >
          <button
            class="button-container is-disabled"
            disabled=""
            type="button"
          >
            <test-file-stub />
          </button>
          <div>
            NaN
          </div>
          <button
            class="button-container"
            type="button"
          >
            <test-file-stub />
          </button>
        </div>
        <div
          class="bonus-display"
        >
          NaN
        </div>
      </div>
      <h2
        class="attribute"
      >
        strength
      </h2>
      <p>
        <span>
          Carry weight:
        </span>
         
        NaNlb
      </p>
      <p>
        <span>
          Push weight:
        </span>
         
        NaNlb
      </p>
      <p>
        <span>
          Long jump:
        </span>
         
        NaN/NaNft
      </p>
      <p>
        <span>
          High jump:
        </span>
         
        NaN/NaNft
      </p>
      <p>
        <span>
          Skills:
        </span>
         
        Athletics
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
